src/pilot/assign_role.py

```diff
@@ -923,8 +926,19 @@ def generate_osd_config(ip_mac_service_tag, drac_client):
 
     # load the osd environment file
     osd_config_file = os.path.join(Constants.TEMPLATES, "ceph-osd-config.yaml")
-    with open(osd_config_file, 'r') as stream:
-        current_osd_configs = yaml.load(stream)
+    stream = open(osd_config_file, 'r')
+    while True:
+        try:
+            fcntl.flock(stream, fcntl.LOCK_EX | fcntl.LOCK_NB)
+            break
+        except IOError as e:
+            if e.errno != errno.EAGAIN:
+                raise
+            else:
+                time.sleep(1)
+    current_osd_configs = yaml.load(stream)
```

Comment:
I think there still is an issue here.  The scenario that I'm thinking of is 1 thread loads the current config and is then swapped out.  Another thread loads the current config, adds a new config to it and saves the file.  When the 1st thread starts back up, it will overwrite the file, losing the entry that the 2nd thread wrote to it.  To avoid this, I think we would have to lock the file for rw and do everything inside the lock in one shot instead of lock, read, release, lock, write.